### PR TITLE
Dotenv becomes toml

### LINKED ISSUE #7274: Consider using something other than dotenv in 8.0.0
(Motivated in part by #7270 and in part by my own experiences.)

I have been thinking about alternatives to dotenv. I have seen these issues with the current scheme:

- The dotenv syntax is poorly specified and isn't consistent between Javascript, Python, shell, and other implementations. (Take a look at the Python dotenv library code.)
- To the beginning user, the syntax is unique, different, and idiosyncratic.
- We don't implement the whole syntax anyway.
- The syntax is hard to parse properly, and uses up more code space than a simpler syntax. Consider the problems we've had with leading and trailing spaces, and `#` comment indicators.
- The `.env` filename with the leading `.` is not handled well by some editors, by Windows and other OS's, and is invisible if hidden file viewing is not turned on in the host OS.

I am thinking of these possibilities:

- Use functionality in `boot.py`, such as `os.environ`, to set some persistent values. Could the web and BLE workflow startups be deferred until after `boot.py` runs?
- Implement a subset of `.ini` or `.toml` syntax: one section only, no nesting.
- Invent our own syntax, maybe something like:
  - Just assignments, without quotes
  - A multiline value would use `\n`. It would not actually be multiline in the file. Or just skip multiline values
  - `\\` escapes `\` if `\n` is used.
  - `# Comment` only permitted on a line by itself.
  - not a `.file` naming scheme

It is not too late to make a change like this: we are still in beta.

### PR DESCRIPTION
**Pull request by jepler**
This PR replaces the `dotenv` module and the `.env` file with the `_environ` module and the `settings.toml` file.

Testing performed:
 * on pico w, set up wifi workflow with alternate port number
 * ran environ_test in build and on pico w

Would appreciate test reports of scenarios I didn't test, like
- [x] ble workflow
- [x] espressif wifi workflow

closes #7274

### CODE REVIEW

File: ports/raspberrypi/common-hal/wifi/Radio.c [hunk 1]
```diff
@@ -215,11 +231,19 @@ wifi_radio_error_t common_hal_wifi_radio_connect(wifi_radio_obj_t *self, uint8_t
...
+    if (ssid_len > 32) {
+        return WIFI_RADIO_ERROR_CONNECTION_FAIL;
+    }
 
     size_t timeout_ms = timeout <= 0 ? 8000 : (size_t)MICROPY_FLOAT_C_FUN(ceil)(timeout * 1000);
     uint64_t start = port_get_raw_ticks(NULL);
     uint64_t deadline = start + timeout_ms;
 
+    if (connection_unchanged(self, ssid, ssid_len)) {
+        mp_printf(&mp_plat_print, "re-used existing wifi connection");
```
Comment (dhalbert) [inline ports/raspberrypi/common-hal/wifi/Radio.c:243]: Is this a debugging printf?

**Review by dhalbert — CHANGES_REQUESTED** [review]
Thank you for taking this on!

I have not yet tested the code, but had some initial questions in the contextual comments. And, do we really need to expose `_environ` as a native module at all? If `os.getenv()` does the same thing, how about making that the sole Python interface?

File: ports/unix/coverage.c [hunk 2]
```diff
@@ -95,6 +95,7 @@ STATIC const mp_rom_map_elem_t rawfile_locals_dict_table[] = {
     { MP_ROM_QSTR(MP_QSTR_readinto), MP_ROM_PTR(&mp_stream_readinto_obj) },
     { MP_ROM_QSTR(MP_QSTR_readline), MP_ROM_PTR(&mp_stream_unbuffered_readline_obj) },
     { MP_ROM_QSTR(MP_QSTR_ioctl), MP_ROM_PTR(&mp_stream_ioctl_obj) },
+    { MP_ROM_QSTR(MP_QSTR_ioctl), MP_ROM_PTR(&mp_stream_ioctl_obj) },
```
Comment (tannewt) [inline ports/unix/coverage.c:98]: Duplicate

File: shared-bindings/os/getenv.c [hunk 3]
```diff
@@ -0,0 +1 @@
+// this file needs to exist but does not have any content
```
Comment (tannewt) [inline shared-bindings/os/getenv.c:1]: Why does this need to exist? Usually files here are either __init__ or a class. This is weird to have for a function.

File: docs/environment.rst [hunk 4]
```diff
@@ -6,24 +6,49 @@ variables are commonly used to store "secrets" such as Wi-Fi passwords and API
...
+
+   str_key="Hello world" # with trailing comment
+   int_key = 7
+   unicode_key="œuvre"
+   unicode_key2="\\u0153uvre" # same as above
+   unicode_key3="\\U00000153uvre" # same as above
+   escape_codes="supported, including \\r\\n\\"\\\\"
+   # comment
+   [subtable]
+   subvalue="cannot retrieve this using _environ or getenv"
```
Comment (dhalbert) [inline docs/environment.rst:35]:
```suggestion
   subvalue="cannot retrieve this using getenv"
```

File: ports/espressif/common-hal/_bleio/Adapter.c [hunk 5]
```diff
@@ -101,22 +101,16 @@ void common_hal_bleio_adapter_set_enabled(bleio_adapter_obj_t *self, bool enable
...
-        char ble_name[32];
-        name_len = dotenv_get_key("/.env", "CIRCUITPY_BLE_NAME", ble_name, sizeof(ble_name) - 1);
-        if (name_len > 0) {
-            if (name_len > MYNEWT_VAL_BLE_SVC_GAP_DEVICE_NAME_MAX_LENGTH) {
-                name_len = MYNEWT_VAL_BLE_SVC_GAP_DEVICE_NAME_MAX_LENGTH;
-            }
-            ble_name[name_len] = '\0';
+        #if CIRCUITPY_OS_GETENV
+        char ble_name[1 + MYNEWT_VAL_BLE_SVC_GAP_DEVICE_NAME_MAX_LENGTH];
+        os_environ_err_t result = common_hal_os_environ_get_key_str("CIRCUITPY_BLE_NAME", ble_name, sizeof(ble_name));
```
Comment (dhalbert) [inline ports/espressif/common-hal/_bleio/Adapter.c:106]: I thought this would now be named for `os.getenv()`?

Similar changes throughout.

```suggestion
        os_getenv_err_t result = common_hal_os_getenv_str("CIRCUITPY_BLE_NAME", ble_name, sizeof(ble_name));
```

File: ports/espressif/supervisor/port.c [hunk 6]
```diff
@@ -519,7 +519,7 @@ void port_idle_until_interrupt(void) {
 void port_post_boot_py(bool heap_valid) {
     if (!heap_valid && filesystem_present()) {
         mp_int_t reserved;
-        if (dotenv_get_key_int("/.env", "CIRCUITPY_RESERVED_PSRAM", &reserved)) {
+        if (common_hal_os_environ_get_key_int("CIRCUITPY_RESERVED_PSRAM", &reserved) == ENVIRON_OK) {
```
Comment (dhalbert) [inline ports/espressif/supervisor/port.c:522]:
```suggestion
        if (common_hal_os_getenv_int("CIRCUITPY_RESERVED_PSRAM", &reserved) == GETENV_OK) {
```

File: ports/atmel-samd/mpconfigport.mk [hunk 7]
```diff
@@ -33,6 +33,7 @@ CIRCUITPY_BLEIO_HCI = 0
 CIRCUITPY_BUILTINS_POW3 ?= 0
 CIRCUITPY_COMPUTED_GOTO_SAVE_SPACE ?= 1
 CIRCUITPY_COUNTIO ?= 0
+CIRCUITPY_OS_GETENV ?= 0
```
Comment (dhalbert) [inline ports/atmel-samd/mpconfigport.mk:36]: Re-alphabetize.

File: ports/unix/main.c [hunk 8]
```diff
@@ -540,6 +547,7 @@ MP_NOINLINE int main_(int argc, char **argv) {
         MP_DECLARE_CONST_FUN_OBJ_0(extra_cpp_coverage_obj);
         mp_store_global(MP_QSTR_extra_coverage, MP_OBJ_FROM_PTR(&extra_coverage_obj));
         mp_store_global(MP_QSTR_extra_cpp_coverage, MP_OBJ_FROM_PTR(&extra_cpp_coverage_obj));
+        mp_store_global(MP_QSTR_get_key, MP_OBJ_FROM_PTR(&get_key_obj));
```
Comment (dhalbert) [inline ports/unix/main.c:550]: Is this right? `get_key` instead of `getenv`?

File: py/circuitpy_defns.mk [hunk 9]
```diff
@@ -589,7 +586,7 @@ SRC_SHARED_MODULE_ALL = \
 	displayio/TileGrid.c \
 	displayio/area.c \
 	displayio/__init__.c \
-	dotenv/__init__.c \
+	_environ/__init__.c \
```
Comment (dhalbert) [inline py/circuitpy_defns.mk:589]: This should be gone?

File: shared-bindings/os/__init__.h [hunk 10]
```diff
@@ -38,6 +38,8 @@ mp_obj_t common_hal_os_uname(void);
 void common_hal_os_chdir(const char *path);
 mp_obj_t common_hal_os_getcwd(void);
 mp_obj_t common_hal_os_getenv(const char *key, mp_obj_t default_);
+mp_obj_t common_hal_os_getenv_path(const char *path, const char *key, mp_obj_t default_);
+
```
Comment (dhalbert) [inline shared-bindings/os/__init__.h:42]: Is `getenv_int` going to be here?

File: shared-module/os/__init__.h [hunk 11]
```diff
@@ -24,13 +24,22 @@
...
+#pragma once
+
+typedef enum {
+    ENVIRON_OK = 0,
+    ENVIRON_ERR_OPEN,
+    ENVIRON_ERR_UNICODE,
+    ENVIRON_ERR_LENGTH,
+    ENVIRON_ERR_NOT_FOUND,
+    ENVIRON_ERR_UNEXPECTED = 0xff00, // logical or'd with the byte value
+} os_environ_err_t;
```
Comment (dhalbert) [inline shared-module/os/__init__.h:36]: For consistency:
```suggestion
typedef enum {
    GETENV_OK = 0,
    GETENV_ERR_OPEN,
    GETENVN_ERR_UNICODE,
    GETENV_ERR_LENGTH,
    GETENV_ERR_NOT_FOUND,
    GETENV_ERR_UNEXPECTED = 0xff00, // logical or'd with the byte value
} os_getenv_err_t;
```

File: shared-module/os/__init__.h [hunk 12]
```diff
@@ -24,13 +24,22 @@
...
+// Allocation free version that returns the full length of the value.
+// If it fits, the return value is 0-terminated. If the value doesn't fit,
+// *value_len may be an over-estimate but never an under-estimate.
+os_environ_err_t common_hal_os_environ_get_key_str(const char *key, char *value, size_t value_len);
 
-// Returns true and sets value to the read value.  Returns false if the value was not numeric.
-bool dotenv_get_key_int(const char *path, const char *key, mp_int_t *value);
+// Returns ENVIRON_ERR_OK and sets value to the read value.  Returns
+// ENVIRON_ERR_... if the value was not numeric. allocation-free.
+os_environ_err_t common_hal_os_environ_get_key_int(const char *key, mp_int_t *value);
```
Comment (dhalbert) [inline shared-module/os/__init__.h:45]: I thought these would be `common_hal_os_getenv...`

File: supervisor/shared/web_workflow/web_workflow.c [hunk 13]
```diff
@@ -244,22 +244,23 @@ void supervisor_web_workflow_status(void) {
...
+    size_t ssid_len = 0;
+    size_t password_len = 0;
 
-    #if CIRCUITPY_DOTENV
-    ssid_len = dotenv_get_key("/.env", "CIRCUITPY_WIFI_SSID", ssid, sizeof(ssid) - 1);
-    password_len = dotenv_get_key("/.env", "CIRCUITPY_WIFI_PASSWORD", password, sizeof(password) - 1);
-    #endif
-    if (ssid_len <= 0 || (size_t)ssid_len >= sizeof(ssid) ||
-        password_len <= 0 || (size_t)password_len >= sizeof(password)) {
+    os_environ_err_t result = common_hal_os_environ_get_key_str("CIRCUITPY_WIFI_SSID", ssid, sizeof(ssid));
```
Comment (dhalbert) [inline supervisor/shared/web_workflow/web_workflow.c:254]: update

File: supervisor/shared/web_workflow/web_workflow.c [hunk 14]
```diff
@@ -244,22 +244,23 @@ void supervisor_web_workflow_status(void) {
...
+    if (result != ENVIRON_OK) {
+        return;
+    }
+
+    result = common_hal_os_environ_get_key_str("CIRCUITPY_WIFI_PASSWORD", password, sizeof(password));
```
Comment (dhalbert) [inline supervisor/shared/web_workflow/web_workflow.c:259]: update

File: supervisor/shared/web_workflow/web_workflow.c [hunk 15]
```diff
@@ -244,22 +244,23 @@ void supervisor_web_workflow_status(void) {
...
+    if (result != ENVIRON_OK) {
```
Comment (dhalbert) [inline supervisor/shared/web_workflow/web_workflow.c:260]:
```suggestion
    if (result != GETENV_OK) {
```

File: supervisor/shared/web_workflow/web_workflow.c [hunk 16]
```diff
@@ -326,11 +318,12 @@ void supervisor_start_web_workflow(void) {
         common_hal_socketpool_socket_listen(&listening, 1);
     }
 
-    mp_int_t api_password_len = dotenv_get_key("/.env", "CIRCUITPY_WEB_API_PASSWORD", _api_password + 1, sizeof(_api_password) - 2);
-    if (api_password_len > 0) {
+
+    const size_t api_password_len = sizeof(_api_password) - 1;
+    result = common_hal_os_environ_get_key_str("CIRCUITPY_WEB_API_PASSWORD", _api_password + 1, api_password_len);
+    if (result == ENVIRON_OK) {
```
Comment (dhalbert) [inline supervisor/shared/web_workflow/web_workflow.c:324]: update

File: tests/circuitpython/environ_test.py [hunk 17]
```diff
@@ -0,0 +1,42 @@
+import os
+
+try:
+    get_key
```
Comment (dhalbert) [inline tests/circuitpython/environ_test.py:4]: Is this the new name for Unix only? Maybe a clearer name for this is `getenv_from_file`?

**Review by dhalbert — CHANGES_REQUESTED** [review]
Thanks for the refactor. I'm wondering if there are some missing checkins here, as the naming doesn't seem to have been redone everywhere. I'm not sure why it compiles.

File: shared-bindings/os/getenv.c (hunk 3 above)
Comment (jepler) [reply shared-bindings/os/getenv.c:1]: It's split out because this implementation (but not the rest of os) needs to be available in the unix port.

but when I added the pattern that would pick up `shared-module/os/getenv.c` during build (in `circuitpy_defns.mk` `SRC_SHARED_MODULE_ALL` variable) builds failed because they refered to this filename. Do I need to list it differently?

**Review by dhalbert — COMMENTED** [review]
I had an idea for simpler testing in the unix port. Instead of adding a special coverage/unix-only method to specify the filename, I think the test program(s) could simply rewrite `settings.toml` over and over again, since each time `os.getenv()` is called it reads the file fresh:
```py
# write settings.toml test values
# do os.getenv() tests
# write settings.toml again with different values
# do more os.getenv() tests
# ...
```

An alternative is to add an extra optional argument to `os.getenv()` that is enabled only for the unix port.

EDIT: .... I see that you do not close the file after reading, which makes the first alternative not work, unless you always close the file. But writing the test files dynamically could save some of the `.toml.bin` pain, maybe.

File: shared-bindings/os/getenv.c (hunk 3 above)
Comment (dhalbert) [reply shared-bindings/os/getenv.c:1]: I think you solved the pattern problem. The function could be moved back to `__init__.c` if the file still poses an issue.

**Comment by dhalbert** [comment]
@jepler On ESP32-S2, setting up web workflow in `settings.toml` is causing a crash very early on. I will debug more after dinner. I can fetch values from the file just fine when the web workflow is not triggered.

**Comment by dhalbert** [comment]
Thanks for 3620d14. Now my Metro ESP32-S2 is coming up in web workflow. I am still have odd crashes. Now I am getting HardFault_Handler safe mode when using the file browser to download a file that is not known to be a text file:
image

I am _not_ having this problem with the tip of `main` right now.
Backtrace is:
```
0x400a8850: _send_str at /home/halbert/repos/circuitpython/ports/espressif/../../supervisor/shared/web_workflow/web_workflow.c:350
0x400a8888: _send_strs at /home/halbert/repos/circuitpython/ports/espressif/../../supervisor/shared/web_workflow/web_workflow.c:360
0x400a8c4d: _reply_with_file at /home/halbert/repos/circuitpython/ports/espressif/../../supervisor/shared/web_workflow/web_workflow.c:694
0x400a9620: _reply at /home/halbert/repos/circuitpython/ports/espressif/../../supervisor/shared/web_workflow/web_workflow.c:1159
0x400a9cee: _process_request at /home/halbert/repos/circuitpython/ports/espressif/../../supervisor/shared/web_workflow/web_workflow.c:1384
0x400a9d18: supervisor_web_workflow_background at /home/halbert/repos/circuitpython/ports/espressif/../../supervisor/shared/web_workflow/web_workflow.c:1402
0x400a2943: workflow_background at /home/halbert/repos/circuitpython/ports/espressif/../../supervisor/shared/workflow.c:55
0x400a17c5: background_callback_run_all at /home/halbert/repos/circuitpython/ports/espressif/../../supervisor/shared/background_callback.c:97
0x400a26cb: supervisor_run_background_tasks_if_tick at /home/halbert/repos/circuitpython/ports/espressif/../../supervisor/shared/tick.c:122
0x400a0fec: run_code_py at /home/halbert/repos/circuitpython/ports/espressif/../../main.c:536
0x400a12e7: main at /home/halbert/repos/circuitpython/ports/espressif/../../main.c:1019
0x400a16df: app_main at /home/halbert/repos/circuitpython/ports/espressif/supervisor/port.c:533
0x40170168: main_task at /home/halbert/repos/circuitpython/ports/espressif/esp-idf/components/freertos/port/port_common.c:141
```
Maybe this is

**Comment by jepler** [comment]
I merged main into this PR a number of times so it should have the startup fix.

**Review by dhalbert — DISMISSED** [review]
Thank you for this, and for patience through all the churn and testing!!

File: shared-bindings/os/getenv.c (hunk 3 above)
Comment (tannewt) [reply shared-bindings/os/getenv.c:1]: You don't need this file if you use SRC_SHARED_MODULE_INTERNAL: https://github.com/adafruit/circuitpython/blob/main/py/circuitpy_defns.mk#L720

I'd put the common_hal_ functions into `__init__.c` in shared-module and the parsing stuff can stay in the "internal file" getenv.c.

**Review by tannewt — CHANGES_REQUESTED** [review]
I realize I'm being nitpicky but I think this reorg is easy to do and makes it more consistent with other modules.

File: shared-bindings/os/getenv.c (hunk 3 above)
Comment (jepler) [reply shared-bindings/os/getenv.c:1]: I will do the 'fix' for needing the nearly-empty file.

Most of the code has to stay in getenv.c so that it can be linked into the unix coverage build for testing.

**Review by tannewt — APPROVED** [review]
Thank you!

**Review by dhalbert — APPROVED** [review]

**Review by dhalbert — APPROVED** [review]
